File: library/src/main/java/com/wangzhen/permission/PermissionManager.java
package com.wangzhen.permission;

import android.content.Context;
import android.content.Intent;
import android.net.Uri;
import android.os.Build;
import android.provider.Settings;

import androidx.fragment.app.Fragment;
import androidx.fragment.app.FragmentActivity;
import androidx.fragment.app.FragmentManager;

import com.wangzhen.permission.callback.PermissionCallback;
import com.wangzhen.permission.callback.PermissionOperate;
import com.wangzhen.permission.fragment.PermissionFragment;
import com.wangzhen.permission.util.Utils;

import static com.wangzhen.permission.common.Common.FRAGMENT_TAG;

/**
 * PermissionManager
 * Created by wangzhen on 2020/4/15.
 */
public final class PermissionManager {
    private static PermissionManager sInstance;
    private static int sCode;

    static {
        sInstance = new PermissionManager();
    }

    public static void request(Fragment fragment, PermissionCallback callback, String... permissions) {
        request(fragment.getActivity(), callback, permissions);
    }

    public static void request(Context context, PermissionCallback callback, String... permissions) {
        request(Utils.getFragmentActivity(context), callback, permissions);
    }

    public static void request(FragmentActivity activity, PermissionCallback callback, String... permissions) {
        if (Build.VERSION.SDK_INT >= Build.VERSION_CODES.M) {
            sInstance.requestPermission(activity, callback, permissions);
        } else {
            if (callback != null) {
                callback.onGrant(permissions);
            }
        }
    }

    /**
     * start permission request
     *
     * @param activity    activity
     * @param callback    callback
     * @param permissions permissions
     */
    private void requestPermission(FragmentActivity activity, PermissionCallback callback, String... permissions) {
        FragmentManager manager = activity.getSupportFragmentManager();
        Fragment tag = manager.findFragmentByTag(FRAGMENT_TAG);
        PermissionFragment fragment;
        if (tag instanceof PermissionOperate) {
            fragment = (PermissionFragment) tag;
        } else {
            fragment = new PermissionFragment();
            manager.beginTransaction().add(fragment, FRAGMENT_TAG).commitAllowingStateLoss();
        }
        ((PermissionOperate) fragment).exeRequestPermissions(permissions, callback, obtainRequestCode());
    }

    /**
     * get intent to application details setting page
     *
     * @param context context
     * @return intent
     */
    public static Intent getSettingIntent(Context context) {
        Intent intent = new Intent(Settings.ACTION_APPLICATION_DETAILS_SETTINGS);
        intent.setData(Uri.parse("package:" + context.getPackageName()));
        return intent;
    }

    /**
     * obtain request code
     * {@link FragmentActivity#checkForValidRequestCode(int)}
     *
     * @return request code
     */
    private static int obtainRequestCode() {
        if ((sCode & 0xffff0000) != 0)
            sCode = 0;
        return sCode++;
    }
}

File: library/src/main/java/com/wangzhen/permission/PermissionManager.kt
package com.wangzhen.permission

import android.content.Context
import android.content.Intent
import android.net.Uri
import android.os.Build
import android.provider.Settings
import androidx.fragment.app.Fragment
import androidx.fragment.app.FragmentActivity
import com.wangzhen.permission.callback.PermissionCallback
import com.wangzhen.permission.callback.PermissionOperate
import com.wangzhen.permission.common.Common.FRAGMENT_TAG
import com.wangzhen.permission.fragment.PermissionFragment
import com.wangzhen.permission.util.Utils.getFragmentActivity

/**
 * PermissionManager
 * Created by wangzhen on 2020/4/15.
 */
object PermissionManager {

    @JvmStatic
    fun request(
        fragment: Fragment?,
        callback: PermissionCallback?,
        vararg permissions: String
    ) {
        request(fragment?.activity, callback, *permissions)
    }

    @JvmStatic
    fun request(context: Context?, callback: PermissionCallback?, vararg permissions: String) {
        context?.let { ctx ->
            request(
                getFragmentActivity(ctx) as FragmentActivity, callback, *permissions
            )
        }
    }

    @JvmStatic
    fun request(
        activity: FragmentActivity,
        callback: PermissionCallback?,
        vararg permissions: String
    ) {
        if (Build.VERSION.SDK_INT >= Build.VERSION_CODES.M) {
            requestPermission(activity, callback, *permissions)
        } else {
            callback?.onGrant(arrayOf(*permissions))
        }
    }

    /**
     * get intent to application details setting page
     *
     * @param context context
     * @return intent
     */
    @JvmStatic
    fun getSettingIntent(context: Context): Intent {
        val intent = Intent(Settings.ACTION_APPLICATION_DETAILS_SETTINGS)
        intent.data = Uri.parse("package:" + context.packageName)
        return intent
    }

    /**
     * start permission request
     *
     * @param activity    activity
     * @param callback    callback
     * @param permissions permissions
     */
    private fun requestPermission(
        activity: FragmentActivity,
        callback: PermissionCallback?,
        vararg permissions: String
    ) {
        val manager = activity.supportFragmentManager
        val tag = manager.findFragmentByTag(FRAGMENT_TAG)
        val fragment: PermissionFragment
        if (tag is PermissionOperate) {
            fragment = tag as PermissionFragment
        } else {
            fragment = PermissionFragment()
            manager.beginTransaction().add(fragment, FRAGMENT_TAG).commitAllowingStateLoss()
        }
        (fragment as PermissionOperate).exeRequestPermissions(arrayOf(*permissions), callback)
    }
}